### PR TITLE
Init libdatrie, libthai, ibus-engines.libthai; build pango with libthai

### LINKED ISSUE #68975: Package Request: ibus-libthai
**Project description**
IBus-LibThai is a Thai input method engine for IBus, based on the LibThai
library.

Currently, it provides 3 keyboard layouts internally:
  - Ketmanee
  - Pattachote
  - TIS-820.2538
(Thai XKB symbols are also supported.)

The input method also filters out invalid input sequences with 3 levels of
strictness:
  - Passthrough (no check)
  - Basic Check
  - Srict Check

For applications that support surrounding text retrieval/deletion,
it can also correct the invalid input sequences.

**Metadata**

* homepage URL: https://linux.thai.net/projects/ibus-libthai
* source URL: 
Git: https://github.com/tlwg/ibus-libthai
FTP: ftp://linux.thai.net/pub/thailinux/software/libthai
* license: GNU General Public License v2.0
* platforms: GNU/Linux

### PR DESCRIPTION
**Pull request by SuperSandro2000**
###### Motivation for this change

Closes #68975

Build pango with libthai support.

###### Things done

This PR includes #102059 to be able to test pango on darwin.

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [x] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/libraries/libdatrie/default.nix [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+  buildInputs = stdenv.lib.optional stdenv.isDarwin libiconv;
+
+  postInstall = ''
+    installManPage man/trietool.1
+  '';
+
+  meta = with stdenv.lib;{
+    homepage = "https://linux.thai.net/~thep/datrie/datrie.html";
+    description = "This is an implementation of double-array structure for representing trie";
+    license = licenses.lgpl21;
```
Comment (jtojnar) [inline pkgs/development/libraries/libdatrie/default.nix:23]: Looks like `lgpl21Plus`:

https://github.com/tlwg/libdatrie/blob/d5d104e22fd49a27aed8a1e9c5ccc836600c05e1/datrie/triedefs.h#L6-L9

File: pkgs/tools/inputmethods/ibus-engines/ibus-libthai/default.nix [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+  src = fetchurl {
+    url = "https://linux.thai.net/pub/ThaiLinux/software/libthai/ibus-libthai-${version}.tar.xz";
+    sha256 = "0iam7308rxkx2xwaabc5wyj7vrxgd4cr95pvwrkm8fr9gh2xnwgv";
+  };
+
+  nativeBuildInputs = [ makeWrapper pkg-config ];
+
+  buildInputs = [ gtk3 ibus libthai ];
+
+  meta = with stdenv.lib; {
```
Comment (jtojnar) [inline pkgs/tools/inputmethods/ibus-engines/ibus-libthai/default.nix:16]: Ibus engines need `meta.isIbusEngine = true;`

File: pkgs/development/libraries/libdatrie/default.nix (hunk 1 above)
Comment (SuperSandro2000) [reply pkgs/development/libraries/libdatrie/default.nix:23]: Maybe we should throw an error if such license is used.

File: pkgs/development/libraries/libthai/default.nix [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+  buildInputs = [ libdatrie ];
+
+  postInstall = ''
+    installManPage man/man3/*.3
+  '';
+
+  meta = with stdenv.lib; {
+    homepage = "https://linux.thai.net/projects/libthai/";
+    description = "Set of Thai language support routines";
+    license = licenses.lgpl21;
```
Comment (jtojnar) [inline pkgs/development/libraries/libthai/default.nix:23]: Same here.

File: pkgs/tools/inputmethods/ibus-engines/ibus-libthai/default.nix [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+    isIbusEngine = true;
+    homepage = "https://linux.thai.net/projects/ibus-libthai";
+    description = "Set of Thai language support routines";
+    license = licenses.lgpl21;
```
Comment (jtojnar) [inline pkgs/tools/inputmethods/ibus-engines/ibus-libthai/default.nix:20]: And here.

File: pkgs/tools/inputmethods/ibus-engines/ibus-libthai/default.nix [hunk 5]
```diff
@@ -0,0 +1,24 @@
+{ stdenv, fetchurl, makeWrapper, pkg-config, ibus, gtk3, libthai }:
+
+stdenv.mkDerivation rec {
+  pname = "libthai";
```
Comment (jtojnar) [inline pkgs/tools/inputmethods/ibus-engines/ibus-libthai/default.nix:4]: ibus-libthai?

File: pkgs/tools/inputmethods/ibus-engines/ibus-libthai/default.nix [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+  };
+
+  nativeBuildInputs = [ makeWrapper pkg-config ];
+
+  buildInputs = [ gtk3 ibus libthai ];
+
+  meta = with stdenv.lib; {
+    isIbusEngine = true;
+    homepage = "https://linux.thai.net/projects/ibus-libthai";
+    description = "Set of Thai language support routines";
```
Comment (jtojnar) [inline pkgs/tools/inputmethods/ibus-engines/ibus-libthai/default.nix:19]: Thai input method engine for IBus

**Comment by jtojnar** [comment]
Rebased this onto nixos-unstable (sans the pango change) and confirmed in a VM that the IM works:

image

**Comment by SuperSandro2000** [comment]
> Rebased this onto nixos-unstable (sans the pango change) and confirmed in a VM that the IM works:

Thanks! I admit that I did not try this which I probably should have done. Also thanks for the reviews.

File: pkgs/tools/inputmethods/ibus-engines/ibus-libthai/default.nix [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+stdenv.mkDerivation rec {
+  pname = "ibus-libthai";
+  version = "0.1.4";
+
+  src = fetchurl {
+    url = "https://linux.thai.net/pub/ThaiLinux/software/libthai/ibus-libthai-${version}.tar.xz";
+    sha256 = "0iam7308rxkx2xwaabc5wyj7vrxgd4cr95pvwrkm8fr9gh2xnwgv";
+  };
+
+  nativeBuildInputs = [ makeWrapper pkg-config ];
```
Comment (jtojnar) [inline pkgs/tools/inputmethods/ibus-engines/ibus-libthai/default.nix:12]: Why is `makeWrapper` here and in the other expressions?

Comment (SuperSandro2000) [reply pkgs/tools/inputmethods/ibus-engines/ibus-libthai/default.nix:12]: I shouldn't package when I am tired. I am not sure.

**Review by jtojnar — APPROVED** [review]
Great job, thanks.

**Pull request merged**